File: python_impl/day01.py

def load_data(data_location = "../inputs/day01.txt"):
    """Loads data from a txt file.
    
    Read lines, strip each one, and return as a list
    """
    with open(data_location, "r") as f:
        data = [q.strip() for q in f.readlines()]
    return data


def part_one(data):
    counter = 0
    dial = 50
    for turn in data:
        # Transform into integer
        # "L50" -> -50
        # "R109" -> 109
        turn = int(turn[1:]) if turn[0] == "R" else int(f"-{turn[1:]}")
        dial = (dial + turn) % 100
        if dial == 0: counter += 1
    return counter


def part_two(data):
    counter = 0
    dial = 50
    for turn in data:
        sign = 1 if turn[0] == "R" else -1
        turn = int(turn[1:])
        for _ in range(turn):
            dial = (dial + sign) % 100
            if dial == 0: counter += 1
    return counter


if __name__ == "__main__":
    data = load_data()
    print(part_one(data.copy()))
    print(part_two(data.copy()))

File: rust_impl/src/day01.rs
use crate::utils;

fn load_data(data_location: &str) -> Vec<i32> {
    let mut ints: Vec<i32> = Vec::new();
    let lines = utils::read_lines(data_location);
    for mut sign in lines {
        let n = sign.split_off(1);
        let mut i_n = n.parse::<i32>().unwrap();
        if sign == "L" {
            i_n *= -1;
        }
        ints.push(i_n);
    }
    return ints;
}

fn part_one(data_location: &str) -> i32 {
    let data: Vec<i32> = load_data(data_location);
    let mut counter: i32 = 0;
    let mut dial: i32 = 50;
    for turn in data {
        dial = (dial + turn) % 100;
        if dial == 0 {
            counter += 1;
        }
    }
    return counter;
}

fn part_two(data_location: &str) -> i32 {
    let data: Vec<i32> = load_data(data_location);
    let mut counter: i32 = 0;
    let mut dial: i32 = 50;
    for turn in data {
        let sign = turn/turn.abs();  // Get just the sign of the turn value
        for _ in 0..turn.abs() {
            dial = (dial + sign) % 100;
            if dial == 0 {
                counter += 1;
            }
        }
    }
    return counter;
}

pub fn run(data_location: &str) {
    println!("{}", part_one(data_location));
    println!("{}", part_two(data_location));
}
